### ISSUE TITLE
Move to ESM

### ISSUE THREAD

**Issue by jeddy3** [stylelint/stylelint#5291]
We'll need to:

1. following the [ESM migration guide](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c):
2. replace lazy import with dynamic imports

---

> ### How can I move my CommonJS project to ESM?

> - Add `"type": "module"` to your package.json.
> - Replace `"main": "index.js"` with `"exports": "./index.js"` in your package.json.
> - Remove `'use strict';` from all JavaScript files.
> - Replace all `require()`/`module.export` with `import`/`export`.
> - Use only full relative file paths for imports: `import x from '.';` → `import x from './index.js';`.
> - If you have a TypeScript type definition (for example, `index.d.ts`), update it to use ESM imports/exports.
> - Optional but recommended, use the [`node:` protocol](https://nodejs.org/api/esm.html#esm_node_imports) for imports.

**Comment by ybiquitous** [comment]
@jeddy3 Should we create a long-lived branch such as `esm`?

**Comment by jeddy3** [comment]
Yep. I've created a [`v14` branch](https://github.com/stylelint/stylelint/tree/v14) for the issues listed in https://github.com/stylelint/stylelint/issues/5205#issue-837034702

> Replace all `require()/module.export` with `import/export`.

We can use https://github.com/knowbee/rona for this. 

> Optional but recommended, use the node: protocol for imports.

Find and replace should suffice for this.

> replace lazy import with dynamic imports

This is probably the trickiest aspect of this issue.

**Comment by hudochenkov** [comment]
Shall we wait with ESM until Jest has non-experimental support for it?

**Comment by jeddy3** [comment]
Let's transpile for Jest for now:
https://bl.ocks.org/rstacruz/511f43265de4939f6ca729a3df7b001c

So that we can update our growing list of dependencies that require ESM.

We can move from transpiling to native ESM Jest when it's available.

**Comment by jeddy3** [comment]
This issue is now unblocked and can be started on.

We can use [putout](https://github.com/coderaiser/putout) to do the bulk of the `import` conversion.

Putting a `.putout.json` config in the root:

```json
{
    "rules": {
        "convert-commonjs-to-esm/require": "on",
        "convert-commonjs-to-esm/exports": "on",
        "convert-commonjs-to-esm/commons": "off",
        "apply-destructuring/array": "off",
        "convert-to-arrow-function": "off",
        "apply-destructuring/object": "off",
        "remove-iife": "off",
        "simplify-ternary": "off",
        "merge-if-statements": "off",
        "apply-optional-chaining": "off",
        "remove-boolean-from-assertions": "off",
        "convert-for-each-to-for-of": "off",
        "regexp/optimize": "off",
        "regexp/remove-useless-regexp": "off",
        "convert-equal-to-strict-equal": "off",
        "regexp/remove-useless-group": "off",
        "convert-array-copy-to-slice": "off",
        "remove-useless-variables/rename": "off",
        "remove-useless-operand": "off",
        "remove-useless-array-constructor": "off",
        "remove-useless-type-convertion/named": "off",
        "convert-assignment-to-comparison": "off",
        "convert-for-to-for-of/length": "off",
        "apply-numeric-separators": "off",
        "remove-useless-variables/remove": "off",
        "remove-console": "off",
        "remove-useless-spread/object": "off",
        "convert-template-to-string": "off",
        "remove-unused-variables": "off",
        "nodejs/convert-promisify-to-fs-promises": "off",
        "remove-useless-await": "off"
    },
    "processors": [
        ["markdown", "off"],
        ["css", "off"],
        ["yaml", "off"],
        ["ignore", "off"]
    ]
}
```

And running:

```
npx putout lib/**/*.js
```

I suggest we tackle this in stages. First, we move to `import` statements. Then let's avoid where we use dynamic requires for points of extension (plugins, shared configs, custom syntax etc) and move rules and formatters to static imports. Once we have this core functionality working, let's investigate dynamic imports for points of extension. Finally, we can look into whether dynamic imports can be used for the built-in rules and formatters to improve performance.

I may have time to get the ball rolling with the move to `import` statements this afternoon. That'll give us a foundation to get a sense of how much work is involved in this move to ESM.

**Comment by hudochenkov** [comment]
Maybe we should wait with this until PostCSS 8 migration is done #4942? The reason is PostCSS 8 has ESM exports, but PostCSS 7 — not.

**Comment by jeddy3** [comment]
Sounds good to me. I'm was dipping into it just now, but I think this and #4942 are going to require more changes to the internals than I originally thought. It's probably best we focus on #4942 first, and then migrate to ESM (to update all ESM-only dependencies).

**Comment by hudochenkov** [comment]
Another important thing that maybe important to postpone ESM migration for a while. With PostCSS 8 there is a big chance, that stylelint plugins should be a little bit different and use PostCSS 8 visitors API. That could be challenging for plugin developers.

If stylelint would be ESM, plugins also have to be ESM, because every plugin imports stylelint. And CommonJS can't import ESM.

Combining this two big changes at the same time could be very problematic for plugin author.

Maybe for v14 let's stay on CommonJS. And in v15 we move to ESM. We could do v15 right after v14. It will allow plugin authors first to migrate to PostCSS 8 and new plugin format. Test it, and then migrate to ESM.

**Comment by ybiquitous** [comment]
FYI. There is a way of ["Conditional exports"](https://nodejs.org/api/packages.html#packages_conditional_exports) for that.

- Provide ESM mainly
- Provide also CJS for backward compatibility

We can generate CJS files via a transpiler like [rollup.js](https://www.rollupjs.org).

**Comment by jeddy3** [comment]
@hudochenkov Those are all good points.

> It will allow plugin authors first to migrate to PostCSS 8 and new plugin format. Test it, and then migrate to ESM.

I originally thought that, in the long run, it'll be better for users if we draw a clean line in the sand:

- new world is ESM and PostCSS 8
- old world is commonjs and PostCSS 7

That'd help users navigate compatible and incompatible plugins and tasks runners. However, I agree that this may be too much to ask of plugin and task runner authors in the short term.

@ybiquitous' suggestion of conditional exports may be the appropriate middle ground. Going pure ESM is probably more suitable for libraries (like the ones we import) than tools like Stylelint at this point. We can take a [dual approach](https://blog.sindresorhus.com/get-ready-for-esm-aa53530b3f77#23d5) for the next year and then see how the dust settles. 

This would allow us to move to ESM ourselves in `14.0.0` and update the 8 of our dependencies that are pure ESM already, including one that patches a vulnerability in its latest release (https://github.com/stylelint/stylelint/issues/5205#issuecomment-834364880).

> We can generate CJS files via a transpiler like rollup.js. (or esbuild)

The good news is that this is a lot more viable now with the removal of syntax (https://github.com/stylelint/stylelint/issues/5289), autoprefixer (https://github.com/stylelint/stylelint/issues/5293) and the `calc` parser (https://github.com/stylelint/stylelint/issues/4731), and by the reduction in the size of the `table` package (https://github.com/gajus/table/pull/147#issue-615931203). I think we can use static imports for rules and formatters now, reserving dynamic imports/requires for points of extension. This would simplify our architecture and reduce the complexity of bundling.

If we decide to dual publish and add the infrastructure to bundle commonjs, then it's sensible to repurpose that infrastructure to bundle ESM. While investigating https://github.com/stylelint/stylelint/issues/2454, I found bundling to give comparable or even better performance than using lazy-import.

It's likely other issues and opportunities will emerge when we begin to more fully explore the move to ESM once all the other `14.0.0` items are complete. At this point, I think we'll be in the best position to decide on the best approach for us.

It's exciting to think what `14.0.0` is shaping up to be; the most modern, secure, sustainable and performant stylelint yet!

**Comment by hudochenkov** [comment]
> It's likely other issues and opportunities will emerge when we begin to more fully explore the move to ESM once all the other 14.0.0 items are complete. At that point, I think we'll be in the best position to decide on the best approach for us.

Yeah, let's see what we'll have at that time. In a good new I have my own plugin, so could get first-hand experience with PostCSS 8 and ESM before final release :) I wish I had smaller plugin :)

> the most modern, secure, sustainable and performant stylelint yet!

Sounds like new iPhone announcement :)

**Comment by jeddy3** [comment]
> In a good new I have my own plugin, so could get first-hand experience with PostCSS 8 and ESM before final release :

Ditto. [`stylelint-scales`](https://github.com/signal-noise/stylelint-scales) isn't as complex as `stylelint-order`, but there are a few rules in it I'll need to migrate.

> Sounds like new iPhone announcement :)

Ha 😃 . Building the excitement :)

**Comment by jeddy3** [comment]
> The good news is that this is a lot more viable now with the removal of syntax (#5289), autoprefixer (#5293) and the calc parser (#4731), and by the reduction in the size of the table package (gajus/table#147 (comment)).

I was curious if this was true. The results of a quick test are promising. It looks like ESBuild can bundle the `v14` branch without issue. I suspect it won't have an issue with ESM either, but the proof is in the pudding.

```console
npx esbuild ./lib/index.js --bundle --outfile=out/index.js --format=cjs --platform=node --minify --metafile=lib.json
```

Will result in a 900kb bundle with 100ms require times. Custom syntaxes and plugins worked fine.

If you like, you can upload the generated metafile to https://www.bundle-buddy.com to see a breakdown. The `v14` branch looks good, with only lodash needing to be optimised.

**Comment by alexander-akait** [comment]
We have big problem with https://github.com/stylelint/stylelint/blob/master/bin/stylelint.js#L6 (`v8-compile-cache`), `import().then()` will not work due https://github.com/zertosh/v8-compile-cache/issues/30, we have the same problem in webpack

**Comment by jeddy3** [comment]
Can we do without the package, especially as we'll probably be bundling in `14.0.0`?

**Comment by alexander-akait** [comment]
> Can we do without the package (at least until it's compatible), especially as we'll probably be bundling in 14.0.0?

It reduces startup time (after first running) around 20-40%, especial for custom plugins/rules/etc

We are working on https://github.com/sokra/node-voo (maybe make sense to add es modules cache support too), in near future I want to migrate webpack (i.e. webpack-cli) on this, so we will get big feedback. 

Honestly, maybe we should continue to be in CJS, and use https://github.com/ai/dual-publish, the ecosystem is not quite ready yet, a lot low-level utils/packages is still on CJS, therefore, in fact, it does not bring any huge benefits. Don't forget many developers (especial boilerplates) have `const stylelint =  require('stylelint');` and migration may take time for them (top level await supported only Node.js v14).  

> top-level await is available "unflagged" in Node.js since v14.8

I personally make no small effort in ESM migrations and really look forward to it.

So Node.js 12 is transitional stage, in this stage low level tools/pages should migrate on ESM, top level projects rewrite code with thoughts future ESM using (i.e. `import()`/`import.meta.url`/etc).

Don't forget about https://github.com/davidtheclark/cosmiconfig/issues/224, JSON files should be read using `fs.readFile`...

Just my onion. Again, I'm not against ESM, I love ESM :star:

**Comment by jeddy3** [comment]
> Honestly, maybe we should continue to be in CJS, and use https://github.com/ai/dual-publish, the ecosystem is not quite ready yet

I think our hand is forced as 8 of our dependencies are ESM-only already. We'll need to weigh up performance costs against keeping our dependencies up to date.

> Don't forget many developers (especial boilerplates) have const `stylelint = require('stylelint');` 

The tentative plan is to author in ESM and dual publish a CJS bundle so that users can still use `stylelint = require('stylelint');`.

There's still a lot of [other `14.0.0` tasks](https://github.com/stylelint/stylelint/issues/5205) to complete before we attempt to migrate to ESM. Once we start the migration, I think it'll be clearer what the compromises are and what's the best way forward is.

**Comment by alexander-akait** [comment]
> The tentative plan is to author in ESM and dual publish a CJS bundle so that users can still use stylelint = require('stylelint');.

Sounds good, with bundler it will be easy

**Comment by jeddy3** [comment]
Current list of ESM-only packages yet to be updated in `v14`:

```
execall            ^2.0.0  →   ^3.0.0     
 get-stdin          ^8.0.0  →   ^9.0.0     
 globby            ^11.0.3  →  ^12.0.2     
 log-symbols        ^4.1.0  →   ^5.0.0     
 mathml-tag-names   ^2.1.3  →   ^3.0.1     
 meow               ^9.0.0  →  ^10.1.1     
 slash              ^3.0.0  →   ^4.0.0     
 string-width       ^4.2.2  →   ^5.0.0     
 strip-ansi         ^6.0.0  →   ^7.0.1 
```

**Comment by ybiquitous** [comment]
In addition, we will need to remove CommonJS globals from the ESLint config (see #5536):

https://github.com/stylelint/stylelint/blob/34bb44b89580dbfbe1f69b39125f8f121ad020d7/package.json#L64-L66

**Comment by jeddy3** [comment]
WIP branch at [refactor-to-esm](https://github.com/stylelint/stylelint/tree/refactor-to-esm).

I used putout to do the bulk of the migration. It did a good job, but quite a few files and imports needed manually fixing up afterwards.

The good news is that:
1. a simple Node.js API use case works
2. simple tests outside of the main engine, e.g. utils, work using `jest-esm-transformer`.

You can pull down the branch and use the following to see the former:

```js
// test.mjs

import { lint } from "./lib/index.mjs";

try {
  const result = await lint({
    config: {
      rules: {
        "color-no-invalid-hex": true,
      },
    },
    code: "a { color: #f }",
    formatter: "string",
  });
  console.log(result.output);
} catch (err) {
  console.error(err.stack);
}
```

The bad news is that not much else is working at the moment:
- bundling with esbuild (either to cjs or sticking with esm)
- tests involving the engine

Both seem to struggle with the use of `simport`.

Jest:

```
FAIL  lib/formatters/__tests__/jsonFormatter.test.js
  ● Test suite failed to run

    Jest encountered an unexpected token

    Jest failed to parse a file. This happens e.g. when your code or its dependencies use non-standard JavaScript syntax, or when Jest is not configured to support such syntax.

    Out of the box Jest supports Babel, which will be used to transform your files into valid JS based on your Babel configuration.

    By default "node_modules" folder is ignored by transformers.

    Here's what you can do:
     • If you are trying to use ECMAScript Modules, see https://jestjs.io/docs/ecmascript-modules for how to enable it.
     • To have some of your "node_modules" files transformed, you can specify a custom "transformIgnorePatterns" in your config.
     • If you need a custom transformation specify a "transform" option in your config.
     • If you simply want to mock your non-JS modules (e.g. binary assets) you can stub them out with the "moduleNameMapper" config option.

    You'll find more details and examples of these config options in the docs:
    https://jestjs.io/docs/configuration
    For information about custom transformations, see:
    https://jestjs.io/docs/code-transformation

    Details:

    /Users/jeddy3/Projects/stylelint/lib/createStylelint.js:32
    } = (0, _simport.createCommons)(import.meta.url);
                                           ^^^^

    SyntaxError: Cannot use 'import.meta' outside a module



      at Runtime.createScriptFromCode (node_modules/jest-runtime/build/index.js:1679:14)
      at Object.<anonymous> (lib/index.js:30:47)
```

(Note that the file is some-what transformed, but not `import.meta`).

And the esbuild bundle on execution errors with:

```
jeddy3@mac stylelint % node test.mjs                                 node:internal/errors:464
    ErrorCaptureStackTrace(err);
    ^

TypeError [ERR_INVALID_ARG_VALUE]: The argument 'filename' must be a file URL object, file URL string, or absolute path string. Received undefined
    at new NodeError (node:internal/errors:371:5)
    at createRequire (node:internal/modules/cjs/loader:1218:11)
    at kn (/Users/jeddy3/Projects/stylelint/out/index.cjs:277:7549)
    at Object.<anonymous> (/Users/jeddy3/Projects/stylelint/out/index.cjs:277:8204)
    at Module._compile (node:internal/modules/cjs/loader:1101:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:196:29)
    at ModuleJob.run (node:internal/modules/esm/module_job:183:25) {
  code: 'ERR_INVALID_ARG_VALUE'
```

Bundling can be done with:

```
npx esbuild ./lib/index.js --bundle --outfile=out/index.cjs --format=cjs --platform=node --minify
npx esbuild ./lib/index.js --bundle --outfile=out/index.mjs --format=esm --platform=node --minify
```

I believe the next step may be to identify an alternative for getting `__dirname` and the like (e.g. importing JSON). Perhaps [this](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c#what-do-i-use-instead-of-__dirname-and-__filename), or something within [this thread](https://github.com/sindresorhus/meta/discussions/15). Then it'll be the case of investigating using dynamic imports for plugins and custom syntaxes. Those two things should remove any instances of `simport`.

It may be that once this is done everything will work fine. However, it's hard to know at this stage.

Feel free to pick up the branch and noodle on it, if it's something you're interested in exploring.

**Comment by ybiquitous** [comment]
@jeddy3 Thanks for the investigation.

If we could avoid using `__dirname`, the problem of `import.meta.url` would be resolved.
Currently, `__dirname` seems to be used for testing:

```console
$ git grep __dirname lib | grep -v __tests__
lib/createStylelint.js:13:const STOP_DIR = IS_TEST ? path.resolve(__dirname, '..') : undefined;
lib/getConfigForFile.js:9:const STOP_DIR = IS_TEST ? path.resolve(__dirname, '..') : undefined;
```

Can we replace it with just `process.cwd()` like this?

```diff
 const IS_TEST = process.env.NODE_ENV === 'test';
-const STOP_DIR = IS_TEST ? path.resolve(__dirname, '..') : undefined;
+const STOP_DIR = IS_TEST ? process.cwd() : undefined;
```

**Comment by jeddy3** [comment]
[Nice list of ESLint rules/plugins for ESM](https://gist.github.com/Jaid/164668c0151ae09d2bc81be78a203dd5).

**Comment by adalinesimonian** [comment]
Following up on https://github.com/stylelint/stylelint/issues/3935#issuecomment-963678973:

In vscode-stylelint, we need support in four key areas to move forward on ESM:

### VS Code Support

> See also: https://github.com/microsoft/vscode/issues/130367, https://github.com/microsoft/vscode/issues/135450

Currently, VS Code does not support ESM extensions, technically. But from what I can gather, that is limited to the entry point, so you can still import an ES module from the entry point using a dynamic `import()`. I was able to get a test to pass that imported an ES module using dynamic imports with `NODE_OPTIONS=--experimental-vm-modules` (needed for Jest to work) but I am as of yet not certain if there are any other potential issues or limitations.

### Jest Support

> See also: https://github.com/facebook/jest/issues/9430, https://github.com/nodejs/node/issues/37648

Jest technically supports ESM experimentally. However, one of the biggest missing features is mocks, and we rely on them heavily in vscode-stylelint, one of the biggest reasons being that we heavily rely on external packages for much of the server/client implementation (e.g. `vscode-languageserver`) and need ways to mock those packages for unit tests.

27.1.1 has an [unstable API for mocks](https://github.com/facebook/jest/issues/9430#issuecomment-915109139) that is severely limited and only supports mocks using factories. Hoisting does not work, so any usage in the project using this unstable API would need to have that file's imports re-written to use top-level await and dynamic imports. Personally, I'd rather wait until there's a more stable API available that covers more ground.

Jest also requires the `--experimental-vm-modules` flag to be set on Node. We can achieve this with VS Code by running it with `NODE_OPTIONS=--experimental-vm-modules`. I was able to run a small test that consisted of an example ES module with an .mjs extension and an example CJS test that loaded the module, which worked fine. There's a caveat — it worked when using JS, but not TypeScript (see below). However, this is a very, very limited test and so I don't know if there are additional roadblocks. That would need further and more thorough investigating.

### TypeScript Support

> See also https://github.com/microsoft/TypeScript/issues/46452

TypeScript 4.4 technically supports ESM, but has some caveats, such as the inability to use .cts/.mts/.cjs/.mjs file properly, and [unwanted transpilation of dynamic import statements](https://github.com/microsoft/TypeScript/issues/43329) when the module type is CommonJS. The TypeScript 4.5 beta release adds this support, as well as support for Node.js 12 and Node.js 14+ ESM environments.

However, it looks like the module resolution support will not make it into the final 4.5 release because of [bugs and concerns with UX](https://github.com/microsoft/TypeScript/issues/46452). Instead, it will ship on nightly releases under an experimental flag. This would mean, though, that we'd still be able to use it if we used TypeScript nightly.

### esbuild Support

> See also: https://github.com/evanw/esbuild/issues/48

As far as I can tell, esbuild support ESM just fine — from everything I can find we won't have any issues here. It would be nice to test this myself to see practically if there would be any issues, but given the concerns above I think that test isn't a high priority at the moment.

### Conclusion

ESM purely in the context of Node.js seems fine to use. But in vscode-stylelint, where we run inside of VS Code and heavily utilize Jest and TypeScript, there are too many surface areas that are rough for me to switch the entire project over to ESM. However, there may be workarounds and ways to load ES modules at runtime. So, if Stylelint itself were to switch to ESM, I think there would be practical ways that we could still interface with it downstream. More investigation and testing is needed beyond the initial investigation I performed here.

**Comment by jeddy3** [comment]
@adalinesimonian Thanks for the detailed write-up of your initial investigation. Unless someone wants to champion the migration to ESM soon, I suspect it'll be a slow burner and hopefully some of these rough edges are resolved in the eco-system by the time we get around to it.

**Comment by ybiquitous** [comment]
Ref: ["Dual CommonJS/ES module packages" on the Node.js documentation](https://nodejs.org/api/packages.html#dual-commonjses-module-packages)

**Comment by jeddy3** [comment]
Let's add [Vitest](https://vitest.dev/) to our investigation. It has a comparable API to Jest, but may offer better support for ESM. I also believe it uses ESBuild under the hood to do any transforms during testing, which would align with our idea of using ESbuild to bundle ESM and CJS versions.

**Comment by hudochenkov** [comment]
PostCSS issue for supporting ESM plugins https://github.com/postcss/postcss/issues/1771 and PR for it https://github.com/postcss/postcss/pull/1773.

**Comment by silverwind** [comment]
> Let's add [Vitest](https://vitest.dev/) to our investigation. It has a comparable API to Jest, but may offer better support for ESM. I also believe it uses ESBuild under the hood to do any transforms during testing, which would align with our idea of using ESbuild to bundle ESM and CJS versions.

+1 to vitest, I think if your package is `type: module`, vitest will just run the module, no transformations involved. Another benefit is that vitest relies on node's native ESM resolution compared to Jest which re-implements it in a buggy way via `jest-resolve`.

Vitest also has the downside that it hard-depends on `vite`, so the bundle size is increased and the existance if the esbuild native dependency is a common source of problems. It seems the vitest authors [do not consider](https://github.com/vitest-dev/vitest/issues/2848) removing the vite dependency so using vitest for node is kind of second-class to them.

**Comment by firefoxic** [comment]
> Vitest also has the downside that it hard-depends on vite, so the package size is increased and the existance of the esbuild native dependency is a common source of problems.

How about `node:test`?
Zero dependencies! 🙃

**Comment by ybiquitous** [comment]
Oh, I didn't know it, but [`node:test` seems available since Node 16.17.0](https://nodejs.org/api/test.html#test-runner). If we could drop versions less than 16.17.0 and there would be no concerns, it would be a leading option. 👍🏼

**Comment by firefoxic** [comment]
I think it will take so long to rewrite all the thousands of tests that we will be able to drop Node < 16.17.0 🤭

**Comment by ybiquitous** [comment]
Um, `node:test` seems to provide a watch mode or coverage support, but still experimental. If we would replace Jest with it, our development experience might be worse. 🤔

**Comment by Mouvedia** [comment]
> I believe the next step may be to identify an alternative for getting `__dirname`…

```js
    import * as url from 'url';
    const __dirname = url.fileURLToPath(new URL('.', import.meta.url));
```

**Comment by silverwind** [comment]
> 

I recommend against such ugly `__dirname` variables. Just do it like this:

```js
// If the API in use accepts `file://` URLs
new URL('../package.json', import.meta.url),

// If the API in use only accepts paths
fileURLToPath(new URL('../package.json', import.meta.url)),
```

**Comment by ybiquitous** [comment]
I'm rewriting tests to ESM since there are no effects on the published code. See:
- #6940
- #6941

Note changing test file extensions from `.js` to `.mjs`. This is required for gradual migration, but I'd like to keep `.js` for non-test files.

**Comment by ybiquitous** [comment]
Currently, I have a migration plan as follows:

1. Rewrite all tests (also scripts) to ESM on the `main` branch.
    - Pros: We can continue the migration without any conflict concerns, keeping compatibility.
    - Cons: We need to rename file extensions from `.js` to `.mjs`, which may make tracking commits history a bit hard.
2. Rewrite published files to ESM on a new branch (maybe `esm`).
    - Keep file extensions (`.js`) for compatibility.
    - Transpile `.js` files (ESM) to `.cjs` files (CJS) via [rollup](https://rollupjs.org/).
    - `package.json` has the `"type": "module" field.
    - `package.json` has the `"exports"` field, including entries for `*.js` and `*.cjs`.

This means Stylelint will provide [Dual packages](https://nodejs.org/api/packages.html#dual-commonjses-module-packages).

Any thoughts? Feel free to comment on the idea if any.

**Comment by silverwind** [comment]
I personally wouldn't do `.mjs` but instead go all-in with `type: module` in package.json, but I see that for sizeable projects a gradual migration to that end state might be preferred.

**Comment by ybiquitous** [comment]
Yes, I'd like to keep `.js` as possible, but now I think it's more important to proceed with the migration gradually.

If we might concern about the inconsistency between `.js` and `.mjs`, we could rename test files to `.js` in the future.

**Comment by ybiquitous** [comment]
> In case you need a helper script for the conversation, i have [this](https://github.com/silverwind/esm.js/blob/master/esm.js?rgh-link-date=2023-06-22T14%3A49%3A47Z), which does like 90% of the work.

Excellent! I'll use it. 👍🏼

**Comment by jaydenseric** [comment]
Please don't move to a dual CJS/ESM package; you will introduce the dual package hazard. You can learn more about the downsides of attempting dual CJS/ESM support here:

https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c?permalink_comment_id=3850849#gistcomment-3850849

Just publish pure ESM.

I highly recommend using the `.mjs` file extension, and not doing package `type` of `"module"`:

- `.mjs` is a very strong signal to humans about the format of the module, it allows you to tell a lot about how you can work with that code just by looking at it in `node_modules` or by seeing the file in `unpkg`. You don't need to go to additional steps of hunting around in `package.json` files for `module` or `type` fields to try to reverse engineer what the intended format is, vs what format your given tool is deriving for that file (which is often different as most people are publishing ESM incorrectly).
- `.mjs` is a very strong signal to tools about the format of the modules. For example, you can easily configure ESLint to enforce ESM or CJS specific rules for the different formats of modules in your project using globs targeting the file extensions.
- There exist certain tools that expect their config files to be named with a `.js` file extension, but they require it as if it were CJS. These tools simply break and become un-configurable when your package has a `type` of `"module"`.

The only time I personally use a package `type` of `"module"` with ESM in `.js` files is when I'm doing project work that has JSX in `.tsx` files, because TypeScript doesn't have `.mtsx` like it does `.mts`, you are forced to only build JSX to ESM in `.js` files.

People are over thinking things and overcomplicating the ecosystem. Publish pure ESM. Configure TypeScript `compilerOptions.module` to `nodenext` and it will hold your hand.

**Comment by ybiquitous** [comment]
@jaydenseric Thanks for the comment! I understood your idea against a dual package.

As you suggested, I would also like to avoid a dual package hazard. However, my biggest concern is *compatibility* within the Stylelint ecosystem.

If `stylelint` were published as a pure ESM package, tons of Stylelint users would face situations like: "Oh, updating Stylelint breaks the CI!" Many Stylelint plugin authors would need to rewrite their packages.

So, I think we need a *compromise* solution with a few major versions, for example:
1. Ship a dual package with the next major version (e.g., v16.0.0)
2. Ship a pure ESM package with the further next major version (e.g., v17.0.0)

This idea is based on the assumption that a dual package hazard would never happen with the `stylelint` package because `stylelint` is not a complex package like a web app.

This is a painful trade-off. 😢

---

About `.mjs`. Your suggestion sounds good. I'm concerned about compatibility with `.js` also here, but it might be better to explicitly switch to `.mjs` as you said.

**Comment by jaydenseric** [comment]
Are you sure it's ok for some plugins to import from the ESM Stylelint modules, and some to import from the CJS ones, at the same time when linting a project occurs? That won't cause problems (other than bloated memory usage etc.)?

You are underestimating the headache it is setting up and maintaining frustrating build tool dependencies to also produce CJS modules.

> Many Stylelint plugin authors would need to rewrite their packages.

That's a good thing. We want the whole Stylelint ecosystem as standard ESM. Then you can take for granted everything will work in browsers, Deno (without npm compatibility mode), and other runtimes that only support standard ESM.

Anyone maintaining a package knows that at some point they need to put in effort to migrate it to ESM; we're not going to use non-standard CJS forever. At some point everything will be ESM. If people are too lazy to migrate, then rightfully the ecosystem needs to migrate away from that package as it's not being maintained.

**Comment by dword-design** [comment]
You could migrate Stylelint to pure ESM and use [jiti](https://www.npmjs.com/package/jiti) to load the plugins and configs. It will be able to load CJS, ESM and TypeScript. [Nuxt](https://www.npmjs.com/package/nuxt) also uses it to load its configs.

**Comment by ybiquitous** [comment]
@dword-design Thanks for sharing `jiti`. Interesting. If you don't mind, could you show an example of how Stylelint users use `jiti` with pure ESM `stylelint` and CJS plugins together?

**Comment by dword-design** [comment]
@ybiquitous Stylelint would load the plugins. Users would write CJS or ESM plugins, then Stylelint would load them like this:

```js
import jiti from 'jiti'

const jitiInstance = jiti(process.cwd(), { interopDefault: true, esmResolve: true })

for (let i = 0; i < plugins.length; ++i) {
  if (typeof plugins[i] === 'string') {
    plugins[i] = jitiInstance(plugins[i])
  }
}
```

**Comment by jaydenseric** [comment]
That wouldn't help people writing CJS Stylelint plugins, because [according to the guide](https://stylelint.io/developer-guide/plugins/#the-anatomy-of-a-plugin) they do this:

```js
const stylelint = require("stylelint");
```

If their developer tooling is correct in their plugin repos, everything will shout at them that you can't require an ESM module.

If in an alternate history Stylelint plugins didn't import from Stylelint, like maybe instead they exported a function that receives the arguments they need to implement a plugin, this wouldn't be a problem. Because Stylelint being pure ESM would be able to import either ESM or CJS plugins and call them with the relevant details.

**Comment by dword-design** [comment]
Ah ok I didn't know that plugins rely on Stylelint. Would probably also question that but that's another story 😋.

**Comment by ybiquitous** [comment]
Thanks for the comments. I now understand the many advantages of ESM.

However, considering the community's migration efforts in the future, I still think we need a major version stage to deprecate CJS and announce the future migration toward ESM.

Revisiting my plan on <https://github.com/stylelint/stylelint/issues/5291#issuecomment-1603455062>:

> 1. Ship a dual package with the next major version (e.g., v16.0.0)
> 2. Ship a pure ESM package with the further next major version (e.g., v17.0.0)

In other words, we should encourage plugin authors to rewrite their packages to ESM between v16.0.0 and v17.0.0.

If we ship both CJS and ESM with the next version (v16.0.0) in the following way, I guess we can minimize the possibility of the dual package hazard:

- No pure ESM dependencies
- The same location of CJS and ESM files, e.g.,
    ```
    lib/
    ├── cli.cjs
    ├── cli.mjs
    ```
- CLI entrypoint (`bin/stylelint.js`) is still CJS
    ```js
    require('../lib/cli')(process.argv.slice(2));
    ```

Any thoughts?

**Comment by Mouvedia** [comment]
> Any thoughts?

Let's add _one_ `console.warning` when the last or first cjs plugin is loaded.

**Comment by Mouvedia** [comment]
https://github.com/stylelint/stylelint/blob/26f9351d733effaf14b9b0aa50cee8fb6060fe96/package.json#L83-L85

These lines will have to be removed.

**Comment by silverwind** [comment]
Would also recommend moving eslint config to it's own file. By putting in in package.json, you are needlessly blowing up the published npm package size.

**Comment by MindTooth** [comment]
I'm sorry that I really don't provide anything new to the discussion.  But just experienced that at least with version 15.10.2, it requires a `.mjs` extension.  I then went on renaming all other config files (eslint, postcss), but then eslint wouldn't work (ref. [new file](https://eslint.org/docs/latest/use/configure/configuration-files-new)).

So, would that mean that I still need to use two different file extensions even with `"type": "module"` which is not desired? 🤔

**Comment by Soitora** [comment]
> But just experienced that at least with version 15.10.2, it requires a `.mjs` extension. I then went on renaming all other config files (eslint, postcss), but then eslint wouldn't work

Try using a `.cjs` extension for now

**Comment by ximex** [comment]
what is missing now to finish this task?

**Comment by ybiquitous** [comment]
We have the following tasks left:

- Migrate `lib/rules/*.js` to ESM
- Add system test cases both for ESM and CJS.

**Comment by ybiquitous** [comment]
## Progress report

I believe the migration to ESM is about to be done. Remaining work:

- #7283
- #7279
- #7280
- Add system test cases for CJS build

**Comment by Mouvedia** [comment]
> Add system test cases for the CJS build

Is this really a blocker?
Another one that we can also postpone, which fixes #7201, is #6836.

**Comment by ybiquitous** [comment]
> Is this really a blocker?

Not a blocker, but it should help for more stable releases.

> Another one that we can also postpone, which fixes #7201, is #6836.

We cannot use pure ESM packages as long as we support both CJS and ESM (v16). Probably, v17 or later might be able to update them.
